### PR TITLE
Trying to improve the Klement-solver

### PR DESCRIPTION
**Pull request by CCsimon123**
Trying to solve https://github.com/SciML/SimpleNonlinearSolve.jl/issues/20 and https://github.com/SciML/SimpleNonlinearSolve.jl/issues/19.
I tried the approach described in https://github.com/SciML/SimpleNonlinearSolve.jl/pull/18

But I have not used the singularity test @YingboMa explained:
"Or ```F = lu(J)```, and the singularity check is then ```abs(F.factors[end, end]) < singular_tol```"
Although that solution was faster I was afraid that the LU-decomposition ```lu(J)``` could result in an error if ```J``` was singular.

Therefore I read about good approaches to checking for a singular matrix and found that
```
cond(J) > tolerance_value
```
was a good approach (https://stackoverflow.com/questions/13145948/how-to-find-out-if-a-matrix-is-singular and https://discourse.julialang.org/t/checking-for-singularity-of-matrix/54746/4)

Is this OK, or do you want me to change something?

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
> Although that solution was faster I was afraid that the LU-decomposition lu(J) could result in an error if J was singular.

Just do `lu(A, check = false)`.

```
  lu(A, pivot = RowMaximum(); check = true) -> F::LU

  Compute the LU factorization of A.

  When check = true, an error is thrown if the decomposition fails. When check = false, responsibility for checking the decomposition's validity (via issuccess) lies with the user.
```

File: src/klement.jl [hunk 1]
```diff
@@ -16,7 +16,7 @@ function SciMLBase.solve(prob::NonlinearProblem,
     x = float(prob.u0)
     fₙ = f(x)
     T = eltype(x)
-    J = ArrayInterfaceCore.zeromatrix(x) + I
+    J = init_J(x)
```
Comment (ChrisRackauckas) [inline src/klement.jl:19]: Do this to Broyden too

**Comment by CCsimon123** [comment]
I didn't manage to get the program non-allocating on static-arrays due to ```@ballocated lu(J, check = false)``` > 0.
To get the program non-allocating for scalars, I split the program into 2 cases.

Just tell me if I can improve anything, 
And thx for all the good feedback, I'm learning a lot from it! :)

File: src/klement.jl [hunk 2]
```diff
@@ -29,33 +29,77 @@ function SciMLBase.solve(prob::NonlinearProblem,
...
-    for _ in 1:maxiters
-        xₙ = xₙ₋₁ - inv(J) * fₙ₋₁
-        fₙ = f(xₙ)
-        Δxₙ = xₙ - xₙ₋₁
-        Δfₙ = fₙ - fₙ₋₁
 
-        # Prevent division by 0
-        denominator = max.(J' .^ 2 * Δxₙ .^ 2, 1e-9)
+    # x is scalar
+    if isa(x, Number)
```
Comment (ChrisRackauckas) [inline src/klement.jl:34]:
```suggestion
    if x isa Number
```

File: src/klement.jl [hunk 3]
```diff
@@ -29,33 +29,77 @@ function SciMLBase.solve(prob::NonlinearProblem,
...
+            J += (k * Δxₙ' .* J) * J
+            F = lu(J, check = false)
 
-        iszero(fₙ) &&
-            return SciMLBase.build_solution(prob, alg, xₙ, fₙ;
-                                            retcode = ReturnCode.Success)
+            # Singularity test
+            if any(abs.(F.U[diagind(F.U)]) .< singular_tol)
+                J = init_J(xₙ)
+                F = lu(J, check = false)
```
Comment (ChrisRackauckas) [inline src/klement.jl:97]: This does  one more LU factorization than is needed. Instead, move the factorization to the top

File: src/klement.jl [hunk 4]
```diff
@@ -29,33 +29,77 @@ function SciMLBase.solve(prob::NonlinearProblem,
...
+
+            k = (Δfₙ - J * Δxₙ) ./ denominator
+            J += (k * Δxₙ' .* J) * J
+            F = lu(J, check = false)
 
-        iszero(fₙ) &&
-            return SciMLBase.build_solution(prob, alg, xₙ, fₙ;
-                                            retcode = ReturnCode.Success)
+            # Singularity test
+            if any(abs.(F.U[diagind(F.U)]) .< singular_tol)
```
Comment (YingboMa) [inline src/klement.jl:95]: This allocate unnecessarily.

Comment (CCsimon123) [reply src/klement.jl:95]: I didn't get ```abs(F.factors[end, end])``` to work for static arrays (```F.factors``` gives error)

Comment (CCsimon123) [reply src/klement.jl:95]: And for non-static arrays ```F.factors[end, end]``` works fine if ```J``` isn't singular.
But if ```J``` is singular, then ```F.factors[end, end]``` didn't give the value that I excpected e.g. ```F.factors[end, end]=3.0``` but ```F.factors[end-1, end-1]=0.0```.

Comment (CCsimon123) [reply src/klement.jl:95]: And for non-static arrays ```abs(F.factors[end, end])``` worked fine if ```J``` was non-singular.
But it didn't work as I excpected when ```J``` was singular, i.e. it gave ```abs(F.factors[end, end])=3.0``` and  ```abs(F.factors[end-1, end-1])=0.0```, resulting in the solver ending up with ```NaN``` as output.

Comment (ChrisRackauckas) [reply src/klement.jl]: Let's open up an issue on this. It may need to be specialized more, but for now we can go with this.

**Pull request merged**